File: matrix/spiral-matrix/spiral-matrix.go
func spiralOrder(matrix [][]int) []int {
    m, n := len(matrix), len(matrix[0])
	res := make([]int, 0, m*n)
	top, bottom := 0, m-1
	left, right := 0, n-1

	for top <= bottom && left <= right {
		for c := left; c <= right; c++ {
			res = append(res, matrix[top][c])
		}
		top++

		for r := top; r <= bottom; r++ {
			res = append(res, matrix[r][right])
		}
		right--

		if top <= bottom {
			for c := right; c >= left; c-- {
				res = append(res, matrix[bottom][c])
			}
			bottom--
		}
		if left <= right {
			for r := bottom; r >= top; r-- {
				res = append(res, matrix[r][left])
			}
			left++
		}
	}
	return res
}

File: matrix/spiral-matrix/spiral-matrix.cs
public class Solution
{
    public IList<int> SpiralOrder(int[][] matrix)
    {
        var res = new List<int>(matrix.Length * matrix[0].Length);
        int top = 0, bottom = matrix.Length - 1;
        int left = 0, right = matrix[0].Length - 1;

        while (top <= bottom && left <= right)
        {
            for (int c = left; c <= right; c++) res.Add(matrix[top][c]);
            top++;

            for (int r = top; r <= bottom; r++) res.Add(matrix[r][right]);
            right--;

            if (top <= bottom)
            {
                for (int c = right; c >= left; c--) res.Add(matrix[bottom][c]);
                bottom--;
            }
            if (left <= right)
            {
                for (int r = bottom; r >= top; r--) res.Add(matrix[r][left]);
                left++;
            }
        }
        return res;
    }
}